Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp

```diff
@@ -50,134 +41,75 @@ SVGResourcesCycleSolver::SVGResourcesCycleSolver(RenderElement& renderer, SVGRes
-
-    // Then operate on the child resources of the given renderer.
-    // <marker id="a"> <path marker-start="url(#b)"/> ...
-    // <marker id="b"> <path marker-start="url(#a)"/> ...
-    for (auto& child : childrenOfType<RenderElement>(renderer)) {
-
-        LOG_DEBUG_CYCLE("(%p) Checking child %p\n", &renderer, &child);
+    ~ActiveResourceAcquisition()
+    {
+        m_set.remove(m_resource);
+    }
 
-        if (auto* childResources = SVGResourcesCache::cachedResourcesForRenderer(child)) {
+private:
+    WeakHashSet<RenderSVGResourceContainer>& m_set;
+    RenderSVGResourceContainer& m_resource;
+};
 
-            LOG_DEBUG_CYCLE("(%p) Child %p had cached resources. Check them.\n", &renderer, &child);
+bool SVGResourcesCycleSolver::resourceContainsCycles(RenderSVGResourceContainer& resource)
+{
+    if (m_acyclicResources.contains(resource))
+        return false;
 
-            // A child of the given 'resource' contains resources.
-            WeakHashSet<RenderSVGResourceContainer> childResourceSet;
-            childResources->buildSetOfResources(childResourceSet);
+    ActiveResourceAcquisition activeResourceAcquisition { m_activeResources, resource };
 
-            // Walk all child resources and check whether they reference any resource contained in the resources set.
-            for (auto& resource : childResourceSet) {
-                LOG_DEBUG_CYCLE("(%p) Child %p had resource %p\n", &renderer, &child, &resource);
-                if (m_allResources.contains(resource))
-                    return true;
+    RenderObject* node = &resource;
+    while (node) {
+        if (node != &resource && node->isSVGResourceContainer()) {
+            node = node->nextInPreOrderAfterChildren(&resource);
+            continue;
+        }
+        if (is<RenderElement>(*node)) {
+            if (auto* resources = SVGResourcesCache::cachedResourcesForRenderer(downcast<RenderElement>(*node))) {
+                WeakHashSet<RenderSVGResourceContainer> resourceSet;
+                resources->buildSetOfResources(resourceSet);
+
+                for (auto& resource : resourceSet) {
+                    if (m_activeResources.contains(resource) || resourceContainsCycles(resource))
+                        return true;
+                }
             }
         }
 
-        LOG_DEBUG_CYCLE("(%p) Recurse into child %p\n", &renderer, &child);
-
-        // Walk children recursively, stop immediately if we found a cycle
-        if (resourceContainsCycles(child))
-            return true;
-
-        LOG_DEBUG_CYCLE("\n(%p) Child %p was ok\n", &renderer, &child);
+        node = node->nextInPreOrder(&resource);
     }
 
-    LOG_DEBUG_CYCLE("\n(%p) No cycles found\n", &renderer);
-
+    m_acyclicResources.add(resource);
     return false;
 }
 
 void SVGResourcesCycleSolver::resolveCycles()
 {
-    ASSERT(m_allResources.isEmptyIgnoringNullReferences());
+    ASSERT(m_activeResources.isEmptyIgnoringNullReferences());
```

Comment:
The life cycle of this class suggests that `m_activeResources` can be just a local variable. So we do not have to check it is empty here and we do not have clear at the end of this function.

Reply:
Yeah, that is quite true. Making this a stain function makes sense to me.